### PR TITLE
[MRG] Gzipping outputs

### LINKED ISSUE #191: the images output by preprocessing pipelines should be gzipped
The current pipeline is very greedy in memory.

### PR DESCRIPTION
**Pull request by mrahim**
Fixes #191 and #202

### CODE REVIEW

**Comment by mrahim** [comment]
The current behavior is that it is compressing all `*.nii` output files at each preprocessing step.

File: pypreprocess/io_utils.py [hunk 1]
```diff
@@ -959,3 +959,47 @@ def sanitize_fwhm(fwhm):
...
+        if not isinstance(ifilename, basestring):
+            raise RuntimeError(
+                "ifilename must be string or list of strings, got %s" % type(
+                    ifilename))
+
+    if not ifilename.endswith('.nii'):
+        return ifilename
+
+    ofilename = ifilename + '.gz'
+    if not output_dir is None:
```
Comment (dohmatob) [inline pypreprocess/io_utils.py]: Change to `if output_dir not is None:`

File: pypreprocess/io_utils.py [hunk 2]
```diff
@@ -959,3 +959,46 @@ def sanitize_fwhm(fwhm):
...
     return fwhm
+
+
+def nii2niigz(ifilename, output_dir=None):
+    """
+    Converts .nii to .nii.gz
+
+    Parameters
+    ----------
+    ifilename: string, of list of strings
```
Comment (bthirion) [inline pypreprocess/io_utils.py]: `input_filename` rather than `ifilename`

File: pypreprocess/io_utils.py [hunk 3]
```diff
@@ -959,3 +959,46 @@ def sanitize_fwhm(fwhm):
...
+        input filename of image to be compressed
+
+    output_dir: string, optional (default None)
+        output directory to which compressed file will be written.
+        If no value is given, the output will be written in the parent
+        directory of ifilename
+
+    Returns
+    -------
+    ofilename: string
```
Comment (bthirion) [inline pypreprocess/io_utils.py]: output_filename

File: pypreprocess/io_utils.py [hunk 4]
```diff
@@ -959,3 +959,46 @@ def sanitize_fwhm(fwhm):
...
+        if not isinstance(ifilename, basestring):
+            raise RuntimeError(
+                "ifilename must be string or list of strings, got %s" % type(
+                    ifilename))
+
+    if not ifilename.endswith('.nii'):
+        return ifilename
+
+    ofilename = ifilename + '.gz'
+    if output_dir is not None:
```
Comment (bthirion) [inline pypreprocess/io_utils.py]: if `output_dir` is `None`, you should compute it as
`output_dir = os.path.dirname(input_filename)`

File: pypreprocess/subject_data.py [hunk 5]
```diff
@@ -446,8 +447,10 @@ def hardlink_output_files(self, final=False):
             if hasattr(self, item):
                 filename = getattr(self, item)
                 if not filename is None:
```
Comment (bthirion) [inline pypreprocess/subject_data.py]: if filename is not None

File: pypreprocess/subject_data.py [hunk 6]
```diff
@@ -461,6 +464,9 @@ def hardlink_output_files(self, final=False):
                 for sess in range(self.n_sessions):
                     filename = filenames[sess]
                     if not filename is None:
```
Comment (bthirion) [inline pypreprocess/subject_data.py]: if filename is not None

File: pypreprocess/subject_data.py [hunk 7]
```diff
@@ -461,6 +464,9 @@ def hardlink_output_files(self, final=False):
...
+                        # gzip before hard link
+                        if item == 'func':
```
Comment (bthirion) [inline pypreprocess/subject_data.py]: why do you condition on item == 'func' ?
This also holds if you use anatomical data.

**Comment by bthirion** [comment]
I'd like to have a test for the new function.
Super useful work, thanks !

File: pypreprocess/subject_data.py (hunk 7 above)
Comment (mrahim) [inline pypreprocess/subject_data.py]: Anatomical data are handled above.
This test is just to avoid taking `realignment_parameters`

**Comment by dohmatob** [comment]
@mrahim: Any updates here ?

**Comment by dohmatob** [comment]
Also, this PR needs rebasing.

**Comment by dohmatob** [comment]
@mrahim: Any progress ?

**Comment by mrahim** [comment]
I need to check the behaviour after rebasing.

**Comment by mrahim** [comment]
It should be okay now. (except for circle ci)

**Comment by dohmatob** [comment]
LGTM. Anyone else ?

**Comment by dohmatob** [comment]
Could you please squash the commits. There are 19+ commits and most have one-word commit messages like "rebasing", "scatch", etc.

File: pypreprocess/subject_data.py [hunk 8]
```diff
@@ -203,6 +208,26 @@ def _sanitize_session_output_dirs(self):
...
+        # session-wise func scratch directories
+        for sess, sess_scratch_dir in enumerate(self.session_scratch_dirs):
+            if sess_scratch_dir is None:
+                if self.n_sessions > 1:
+                    sess_scratch_dir = os.path.join(
+                        self.scratch, self.session_ids[sess])
+                else:
+                    sess_scratch_dir = self.scratch
+            else:
+                sess_scratch_dir = self.scratch
```
Comment (bthirion) [inline pypreprocess/subject_data.py]: I'm not sure I understand the logic here: if `sess_scracth_dir` is not None, it gets overriden ?

**Comment by bthirion** [comment]
LGTM otherwise.

File: pypreprocess/subject_data.py (hunk 8 above)
Comment (mrahim) [inline pypreprocess/subject_data.py]: +1, `sess_scratch_dir` should remain the same if `not None`.

File: pypreprocess/subject_data.py [hunk 9]
```diff
@@ -141,6 +142,10 @@ def __init__(self, func=None, anat=None, subject_id="sub001",
         self.nipype_results = {}
         self._set_items(**kwargs)
         self.scratch = output_dir if scratch is None else scratch
+        self.anat_scratch_dir = anat_output_dir if scratch is None else scratch
+        self.session_scratch_dirs = \
```
Comment (dohmatob) [inline pypreprocess/subject_data.py]: Avoid using "\". You can break before an if or and else, and then you don't need the abnoxious "\".

File: pypreprocess/subject_data.py [hunk 10]
```diff
@@ -141,6 +142,10 @@ def __init__(self, func=None, anat=None, subject_id="sub001",
...
+            session_output_dirs if scratch is None \
+            else [scratch]*len(session_output_dirs)
```
Comment (dohmatob) [inline pypreprocess/subject_data.py]: Missing spaces around "*" operator.

**Comment by dohmatob** [comment]
@mrahim: do you confirm this is this ripe for merge ?

**Comment by mrahim** [comment]
Yes

**Pull request merged**